### PR TITLE
Add pre-upgrade check to test cluster routing allocation is enabled

### LINKED ISSUE #39339: Internal index unrecoverable when upgrade from 5.6.x fails to reindex
When following the steps mentioned in the [upgrade guide](https://www.elastic.co/guide/en/elastic-stack/6.6/upgrading-elastic-stack.html),
if we disable the cluster shard allocation but fail to enable it after
upgrading the nodes and plugins, the next step of upgrading internal
indices fails. As we did not check the bulk response for reindexing,
we delete the old index assuming it has been reindexed successfully.
This is fatal as we cannot recover from this state since index to be upgraded
has been deleted.

**Steps to follow:-**
Following the upgrade guide:
https://www.elastic.co/guide/en/elastic-stack/6.6/upgrading-elastic-stack.html

- Before the upgrade, we disable cluster shard allocation (Step 1)
- And after the upgrade of the node(s) to 6.6, plugins upgrade we need to enable the cluster shard allocation (Step 8 )
- And then run the upgrade internal indices step if you are upgrading from 5.6 (https://www.elastic.co/guide/en/elastic-stack/6.6/upgrading-elastic-stack.html#upgrade-internal-indices)

If we miss the step of enabling cluster shard allocation, during reindex operation `UnavailableShardsException` is thrown.
But as we do not check the `BulkByScrollResponse` if there were any failures and proceed as a successful operation thereby deleting the internal index.

**Solution:-**
There should be a pre-upgrade check before proceeding with the upgrade. Also, in the case where reindex fails for some reason, it should not delete the original index.

### PR DESCRIPTION
**Pull request by bizybot**
When following the steps mentioned in upgrade guide
https://www.elastic.co/guide/en/elastic-stack/6.6/upgrading-elastic-stack.html
if we disable the cluster shard allocation but fail to enable it after
upgrading the nodes and plugins, the next step of upgrading internal
indices fails. As we did not check the bulk request response for reindexing,
we delete the old index assuming it has been created. This is fatal
as we cannot recover from this state.

This commit adds a pre-upgrade check to test the cluster shard
allocation setting and fail upgrade if it is disabled. In case there
are search or bulk failures then we remove the read-only block and
fail the upgrade index request.

Closes #39339

### CODE REVIEW

**Review by albertzaharovits — APPROVED** [review]
LGTM! See if you can find someone from core-infra-features to also have a look.

File: x-pack/plugin/upgrade/src/main/java/org/elasticsearch/xpack/upgrade/IndexUpgradeCheck.java [hunk 1]
```diff
@@ -106,4 +118,9 @@ public void upgrade(TaskId task, IndexMetaData indexMetaData, ClusterState state
                         ActionListener<BulkByScrollResponse> listener) {
         reindexer.upgrade(task, indexMetaData.getIndex().getName(), state, listener);
     }
+
+    // pkg scope for testing
+    InternalIndexReindexer getInternalIndexReindexer() {
```
Comment (imotov) [inline x-pack/plugin/upgrade/src/main/java/org/elasticsearch/xpack/upgrade/IndexUpgradeCheck.java]: That concerns me a bit. Could you add a comment why this is necessary?

**Review by imotov — APPROVED** [review]
LGTM. Left one comment regarding adding testing-only method.

**Comment by tvernum** [comment]
@bizybot I think this should be backported to 5.6 as well. We recommend that users run the `/_xpack/migration/upgrade` while they're on 5.6, so it is important to fix it there.

**Pull request merged**